File: MiniDB_Projects/Team_QuorumDB/src/minidb/sql/parser.py
"""A hand-written tokenizer + recursive-descent parser for MiniDB SQL.

Small enough to read top-to-bottom, it produces the AST nodes in ``ast.py``.
SQL keywords are case-insensitive; identifiers are returned as written.
"""

from __future__ import annotations

from typing import Any, List, NamedTuple, Optional

from . import ast

_KEYWORDS = {
    "CREATE", "TABLE", "INDEX", "ON", "UNIQUE", "DROP", "PRIMARY", "KEY",
    "NOT", "NULL", "INT", "INTEGER", "FLOAT", "DOUBLE", "TEXT", "VARCHAR",
    "BOOL", "BOOLEAN", "INSERT", "INTO", "VALUES", "DELETE", "FROM", "SELECT",
    "WHERE", "AND", "JOIN", "INNER", "BEGIN", "COMMIT", "ROLLBACK", "ABORT",
    "TRUE", "FALSE", "AS",
}
_TYPE_KEYWORDS = {"INT": "INT", "INTEGER": "INT", "FLOAT": "FLOAT",
                  "DOUBLE": "FLOAT", "TEXT": "TEXT", "VARCHAR": "TEXT",
                  "BOOL": "BOOL", "BOOLEAN": "BOOL"}


class ParseError(Exception):
    pass


class Token(NamedTuple):
    kind: str          # 'kw' | 'ident' | 'num' | 'str' | 'op' | 'punct' | 'eof'
    value: Any


def tokenize(sql: str) -> List[Token]:
    tokens: List[Token] = []
    i, n = 0, len(sql)
    while i < n:
        c = sql[i]
        if c.isspace():
            i += 1
            continue
        if c == "-" and i + 1 < n and sql[i + 1] == "-":   # line comment
            while i < n and sql[i] != "\n":
                i += 1
            continue
        if c in "().,;*":
            tokens.append(Token("punct", c))
            i += 1
            continue
        if c in "=<>!":
            two = sql[i:i + 2]
            if two in ("<=", ">=", "<>", "!="):
                tokens.append(Token("op", "!=" if two in ("<>", "!=") else two))
                i += 2
            elif c == "=":
                tokens.append(Token("op", "="))
                i += 1
            elif c in "<>":
                tokens.append(Token("op", c))
                i += 1
            else:
                raise ParseError(f"unexpected character {c!r}")
            continue
        if c in "'\"":                                     # string literal
            i += 1
            start = i
            buf = []
            while i < n and sql[i] != c:
                buf.append(sql[i])
                i += 1
            if i >= n:
                raise ParseError("unterminated string literal")
            i += 1
            tokens.append(Token("str", "".join(buf)))
            continue
        if c.isdigit() or (c == "-" and i + 1 < n and sql[i + 1].isdigit()):
            start = i
            i += 1
            is_float = False
            while i < n and (sql[i].isdigit() or sql[i] == "."):
                if sql[i] == ".":
                    is_float = True
                i += 1
            text = sql[start:i]
            tokens.append(Token("num", float(text) if is_float else int(text)))
            continue
        if c.isalpha() or c == "_":
            start = i
            i += 1
            while i < n and (sql[i].isalnum() or sql[i] == "_"):
                i += 1
            word = sql[start:i]
            up = word.upper()
            tokens.append(Token("kw", up) if up in _KEYWORDS else Token("ident", word))
            continue
        raise ParseError(f"unexpected character {c!r}")
    tokens.append(Token("eof", None))
    return tokens


class Parser:
    def __init__(self, sql: str):
        self.tokens = tokenize(sql)
        self.pos = 0

    # -- token helpers ------------------------------------------------------
    def _peek(self) -> Token:
        return self.tokens[self.pos]

    def _next(self) -> Token:
        tok = self.tokens[self.pos]
        self.pos += 1
        return tok

    def _accept_kw(self, *kw: str) -> bool:
        tok = self._peek()
        if tok.kind == "kw" and tok.value in kw:
            self.pos += 1
            return True
        return False

    def _expect_kw(self, *kw: str) -> str:
        tok = self._next()
        if tok.kind != "kw" or tok.value not in kw:
            raise ParseError(f"expected {' or '.join(kw)}, got {tok.value!r}")
        return tok.value

    def _expect_punct(self, ch: str) -> None:
        tok = self._next()
        if tok.kind != "punct" or tok.value != ch:
            raise ParseError(f"expected {ch!r}, got {tok.value!r}")

    def _accept_punct(self, ch: str) -> bool:
        tok = self._peek()
        if tok.kind == "punct" and tok.value == ch:
            self.pos += 1
            return True
        return False

    def _expect_ident(self) -> str:
        tok = self._next()
        if tok.kind != "ident":
            raise ParseError(f"expected identifier, got {tok.value!r}")
        return tok.value

    # -- entry points -------------------------------------------------------
    def parse(self):
        stmt = self._statement()
        self._accept_punct(";")
        if self._peek().kind != "eof":
            raise ParseError("trailing tokens after statement")
        return stmt

    @staticmethod
    def parse_script(sql: str) -> List[Any]:
        """Split a string of ;-separated statements and parse each."""
        out = []
        for chunk in _split_statements(sql):
            if chunk.strip():
                out.append(Parser(chunk).parse())
        return out

    def _statement(self):
        tok = self._peek()
        if tok.kind != "kw":
            raise ParseError(f"expected a statement, got {tok.value!r}")
        kw = tok.value
        if kw == "CREATE":
            return self._create()
        if kw == "DROP":
            self._next(); self._expect_kw("TABLE")
            return ast.DropTable(self._expect_ident())
        if kw == "INSERT":
            return self._insert()
        if kw == "DELETE":
            return self._delete()
        if kw == "SELECT":
            return self._select()
        if kw == "BEGIN":
            self._next(); return ast.Begin()
        if kw == "COMMIT":
            self._next(); return ast.Commit()
        if kw in ("ROLLBACK", "ABORT"):
            self._next(); return ast.Rollback()
        raise ParseError(f"unsupported statement: {kw}")

    # -- DDL ----------------------------------------------------------------
    def _create(self):
        self._expect_kw("CREATE")
        if self._accept_kw("TABLE"):
            return self._create_table()
        unique = self._accept_kw("UNIQUE")
        self._expect_kw("INDEX")
        name = None
        if self._peek().kind == "ident":
            name = self._expect_ident()
        self._expect_kw("ON")
        table = self._expect_ident()
        self._expect_punct("(")
        column = self._expect_ident()
        self._expect_punct(")")
        return ast.CreateIndex(table=table, column=column, unique=unique, name=name)

    def _create_table(self):
        name = self._expect_ident()
        self._expect_punct("(")
        columns: List[ast.ColumnDef] = []
        pk: Optional[str] = None
        while True:
            col = self._expect_ident()
            type_tok = self._next()
            if type_tok.kind != "kw" or type_tok.value not in _TYPE_KEYWORDS:
                raise ParseError(f"expected a column type, got {type_tok.value!r}")
            col_type = _TYPE_KEYWORDS[type_tok.value]
            # optional VARCHAR(n) length — accepted and ignored
            if self._accept_punct("("):
                self._next()
                self._expect_punct(")")
            not_null = False
            is_pk = False
            while True:
                if self._accept_kw("PRIMARY"):
                    self._expect_kw("KEY")
                    is_pk = True
                    not_null = True
                elif self._accept_kw("NOT"):
                    self._expect_kw("NULL")
                    not_null = True
                else:
                    break
            if is_pk:
                pk = col
            columns.append(ast.ColumnDef(col, col_type, not_null=not_null,
                                         primary_key=is_pk))
            if self._accept_punct(","):
                continue
            break
        self._expect_punct(")")
        return ast.CreateTable(name=name, columns=columns, pk_column=pk)

    # -- DML ----------------------------------------------------------------
    def _insert(self):
        self._expect_kw("INSERT")
        self._expect_kw("INTO")
        table = self._expect_ident()
        columns: Optional[List[str]] = None
        if self._accept_punct("("):
            columns = [self._expect_ident()]
            while self._accept_punct(","):
                columns.append(self._expect_ident())
            self._expect_punct(")")
        self._expect_kw("VALUES")
        rows: List[List[Any]] = []
        while True:
            self._expect_punct("(")
            values = [self._value()]
            while self._accept_punct(","):
                values.append(self._value())
            self._expect_punct(")")
            rows.append(values)
            if self._accept_punct(","):
                continue
            break
        return ast.Insert(table=table, columns=columns, rows=rows)

    def _delete(self):
        self._expect_kw("DELETE")
        self._expect_kw("FROM")
        table = self._expect_ident()
        where = self._where()
        return ast.Delete(table=table, where=where)

    def _select(self):
        self._expect_kw("SELECT")
        columns: List[str] = []
        if self._accept_punct("*"):
            columns = ["*"]
        else:
            columns.append(self._column_name())
            while self._accept_punct(","):
                columns.append(self._column_name())
        self._expect_kw("FROM")
        from_table = self._expect_ident()
        from_alias = self._opt_alias()
        joins: List[ast.JoinClause] = []
        while True:
            self._accept_kw("INNER")
            if not self._accept_kw("JOIN"):
                break
            jtable = self._expect_ident()
            jalias = self._opt_alias()
            self._expect_kw("ON")
            on = self._comparison()
            if not on.is_join_predicate:
                raise ParseError("JOIN ON must compare two columns")
            joins.append(ast.JoinClause(table=jtable, on=on, alias=jalias))
        where = self._where()
        return ast.Select(columns=columns, from_table=from_table,
                          from_alias=from_alias, joins=joins, where=where)

    def _opt_alias(self) -> Optional[str]:
        """An optional table alias: 'AS x' or a bare identifier."""
        if self._accept_kw("AS"):
            return self._expect_ident()
        if self._peek().kind == "ident":
            return self._next().value
        return None

    # -- shared fragments ---------------------------------------------------
    def _column_name(self) -> str:
        """A possibly-qualified column reference, returned as written."""
        first = self._expect_ident()
        if self._accept_punct("."):
            second = self._expect_ident()
            return f"{first}.{second}"
        return first

    def _column_ref(self) -> ast.ColumnRef:
        first = self._expect_ident()
        if self._accept_punct("."):
            second = self._expect_ident()
            return ast.ColumnRef(name=second, table=first)
        return ast.ColumnRef(name=first)

    def _where(self) -> ast.Predicate:
        pred = ast.Predicate()
        if self._accept_kw("WHERE"):
            pred.comparisons.append(self._comparison())
            while self._accept_kw("AND"):
                pred.comparisons.append(self._comparison())
        return pred

    def _comparison(self) -> ast.Comparison:
        left = self._column_ref()
        op_tok = self._next()
        if op_tok.kind != "op":
            raise ParseError(f"expected a comparison operator, got {op_tok.value!r}")
        # right side: another column or a literal
        nxt = self._peek()
        if nxt.kind == "ident":
            right: Any = self._column_ref()
        else:
            right = ast.Literal(self._value())
        return ast.Comparison(left=left, op=op_tok.value, right=right)

    def _value(self) -> Any:
        tok = self._next()
        if tok.kind in ("num", "str"):
            return tok.value
        if tok.kind == "kw":
            if tok.value == "TRUE":
                return True
            if tok.value == "FALSE":
                return False
            if tok.value == "NULL":
                return None
        raise ParseError(f"expected a literal value, got {tok.value!r}")


def _split_statements(sql: str) -> List[str]:
    """Split on semicolons that are not inside string literals."""
    out, buf, quote = [], [], None
    for ch in sql:
        if quote:
            buf.append(ch)
            if ch == quote:
                quote = None
        elif ch in "'\"":
            quote = ch
            buf.append(ch)
        elif ch == ";":
            out.append("".join(buf))
            buf = []
        else:
            buf.append(ch)
    if buf:
        out.append("".join(buf))
    return out


def parse(sql: str):
    return Parser(sql).parse()

File: MiniDB_Projects/Team_minions/src/query/parser.cpp
#include "minidb/query/parser.h"

#include <cctype>
#include <set>

#include "minidb/exceptions.h"

namespace minidb {

namespace {

const std::set<std::string>& keywords() {
    static const std::set<std::string> kw = {
        "CREATE", "TABLE",  "INDEX",  "ON",     "INSERT", "INTO",
        "VALUES", "SELECT", "FROM",   "WHERE",  "JOIN",   "INNER",
        "DELETE", "BEGIN",  "COMMIT", "ABORT",  "ROLLBACK", "AND",
        "PRIMARY","KEY",    "INT",    "INTEGER","TEXT",   "AS",
        "TRANSACTION"};
    return kw;
}

std::string to_upper(const std::string& s) {
    std::string r = s;
    for (char& c : r) c = static_cast<char>(std::toupper((unsigned char)c));
    return r;
}

std::vector<Token> tokenize(const std::string& sql) {
    std::vector<Token> toks;
    std::size_t i = 0, n = sql.size();
    while (i < n) {
        char c = sql[i];
        if (std::isspace((unsigned char)c)) { ++i; continue; }

        if (std::isalpha((unsigned char)c) || c == '_') {
            std::size_t j = i;
            while (j < n && (std::isalnum((unsigned char)sql[j]) || sql[j] == '_'))
                ++j;
            std::string word = sql.substr(i, j - i);
            std::string up = to_upper(word);
            if (keywords().count(up)) {
                toks.push_back({TokKind::KEYWORD, up, 0});
            } else {
                toks.push_back({TokKind::IDENT, word, 0});
            }
            i = j;
            continue;
        }
        if (std::isdigit((unsigned char)c)) {
            std::size_t j = i;
            while (j < n && std::isdigit((unsigned char)sql[j])) ++j;
            Token t{TokKind::NUMBER, sql.substr(i, j - i), 0};
            t.number = std::stoll(t.text);
            toks.push_back(t);
            i = j;
            continue;
        }
        if (c == '\'') {
            std::size_t j = i + 1;
            std::string val;
            while (j < n && sql[j] != '\'') val.push_back(sql[j++]);
            if (j >= n) throw SQLException("unterminated string literal");
            toks.push_back({TokKind::STRING, val, 0});
            i = j + 1;
            continue;
        }
        // Multi-character operators first.
        auto two = [&](const char* op) {
            return i + 1 < n && sql[i] == op[0] && sql[i + 1] == op[1];
        };
        if (two("!=")) { toks.push_back({TokKind::SYMBOL, "!=", 0}); i += 2; continue; }
        if (two("<>")) { toks.push_back({TokKind::SYMBOL, "!=", 0}); i += 2; continue; }
        if (two("<=")) { toks.push_back({TokKind::SYMBOL, "<=", 0}); i += 2; continue; }
        if (two(">=")) { toks.push_back({TokKind::SYMBOL, ">=", 0}); i += 2; continue; }

        std::string sym(1, c);
        if (std::string("()*,.;=<>-").find(c) != std::string::npos) {
            toks.push_back({TokKind::SYMBOL, sym, 0});
            ++i;
            continue;
        }
        throw SQLException(std::string("unexpected character '") + c + "'");
    }
    toks.push_back({TokKind::END, "", 0});
    return toks;
}

CompOp symbol_to_op(const std::string& s) {
    if (s == "=") return CompOp::EQ;
    if (s == "!=") return CompOp::NE;
    if (s == "<") return CompOp::LT;
    if (s == "<=") return CompOp::LE;
    if (s == ">") return CompOp::GT;
    if (s == ">=") return CompOp::GE;
    throw SQLException("expected a comparison operator, got '" + s + "'");
}

}  // namespace

// --- token helpers ----------------------------------------------------------
bool Parser::is_keyword(const std::string& kw) const {
    return peek().kind == TokKind::KEYWORD && peek().text == kw;
}
bool Parser::is_symbol(const std::string& s) const {
    return peek().kind == TokKind::SYMBOL && peek().text == s;
}
bool Parser::accept_keyword(const std::string& kw) {
    if (is_keyword(kw)) { advance(); return true; }
    return false;
}
bool Parser::accept_symbol(const std::string& s) {
    if (is_symbol(s)) { advance(); return true; }
    return false;
}
void Parser::expect_symbol(const std::string& s) {
    if (!accept_symbol(s)) throw SQLException("expected '" + s + "'");
}
void Parser::expect_keyword(const std::string& kw) {
    if (!accept_keyword(kw)) throw SQLException("expected keyword " + kw);
}
std::string Parser::expect_ident() {
    if (peek().kind != TokKind::IDENT)
        throw SQLException("expected an identifier, got '" + peek().text + "'");
    return advance().text;
}

// --- entry point ------------------------------------------------------------
Statement Parser::parse(const std::string& sql) {
    Parser p(tokenize(sql));
    Statement s = p.parse_statement();
    p.accept_symbol(";");
    if (p.peek().kind != TokKind::END)
        throw SQLException("unexpected tokens after statement");
    return s;
}

Statement Parser::parse_statement() {
    if (is_keyword("CREATE")) return parse_create();
    if (is_keyword("INSERT")) return parse_insert();
    if (is_keyword("SELECT")) return parse_select();
    if (is_keyword("DELETE")) return parse_delete();
    if (accept_keyword("BEGIN")) {
        accept_keyword("TRANSACTION");
        return Statement{StmtType::BEGIN, {}, {}, {}, {}, {}};
    }
    if (accept_keyword("COMMIT"))
        return Statement{StmtType::COMMIT, {}, {}, {}, {}, {}};
    if (accept_keyword("ABORT") || accept_keyword("ROLLBACK"))
        return Statement{StmtType::ABORT, {}, {}, {}, {}, {}};
    throw SQLException("unrecognised statement starting with '" +
                       peek().text + "'");
}

Statement Parser::parse_create() {
    expect_keyword("CREATE");
    Statement s;
    if (accept_keyword("TABLE")) {
        s.type = StmtType::CREATE_TABLE;
        s.create_table.table = expect_ident();
        expect_symbol("(");
        do {
            ColumnDef col;
            col.name = expect_ident();
            if (accept_keyword("INT") || accept_keyword("INTEGER"))
                col.type = Type::INT;
            else if (accept_keyword("TEXT"))
                col.type = Type::TEXT;
            else
                throw SQLException("expected a column type (INT or TEXT)");
            if (accept_keyword("PRIMARY")) {
                expect_keyword("KEY");
                col.primary_key = true;
            }
            s.create_table.columns.push_back(col);
        } while (accept_symbol(","));
        expect_symbol(")");
        return s;
    }
    if (accept_keyword("INDEX")) {
        s.type = StmtType::CREATE_INDEX;
        s.create_index.index_name = expect_ident();
        expect_keyword("ON");
        s.create_index.table = expect_ident();
        expect_symbol("(");
        s.create_index.column = expect_ident();
        expect_symbol(")");
        return s;
    }
    throw SQLException("expected TABLE or INDEX after CREATE");
}

Statement Parser::parse_insert() {
    expect_keyword("INSERT");
    expect_keyword("INTO");
    Statement s;
    s.type = StmtType::INSERT;
    s.insert.table = expect_ident();
    if (accept_symbol("(")) {
        do {
            s.insert.columns.push_back(expect_ident());
        } while (accept_symbol(","));
        expect_symbol(")");
    }
    expect_keyword("VALUES");
    do {
        expect_symbol("(");
        std::vector<Value> row;
        do {
            row.push_back(parse_value());
        } while (accept_symbol(","));
        expect_symbol(")");
        s.insert.rows.push_back(std::move(row));
    } while (accept_symbol(","));
    return s;
}

Statement Parser::parse_select() {
    expect_keyword("SELECT");
    Statement s;
    s.type = StmtType::SELECT;
    if (accept_symbol("*")) {
        s.select.select_star = true;
    } else {
        do {
            s.select.columns.push_back(parse_column_ref());
        } while (accept_symbol(","));
    }
    expect_keyword("FROM");
    s.select.from_table = expect_ident();
    // Optional table alias: "FROM t a" or "FROM t AS a".
    if (accept_keyword("AS")) {
        s.select.from_alias = expect_ident();
    } else if (peek().kind == TokKind::IDENT) {
        s.select.from_alias = advance().text;
    }
    // Zero or more joins.
    while (is_keyword("JOIN") || is_keyword("INNER")) {
        accept_keyword("INNER");
        expect_keyword("JOIN");
        JoinClause j;
        j.table = expect_ident();
        if (accept_keyword("AS")) {
            j.alias = expect_ident();
        } else if (peek().kind == TokKind::IDENT) {
            j.alias = advance().text;
        }
        expect_keyword("ON");
        j.on = parse_predicate();
        s.select.joins.push_back(j);
    }
    if (is_keyword("WHERE")) s.select.where = parse_where();
    return s;
}

Statement Parser::parse_delete() {
    expect_keyword("DELETE");
    expect_keyword("FROM");
    Statement s;
    s.type = StmtType::DELETE;
    s.del.table = expect_ident();
    if (is_keyword("WHERE")) s.del.where = parse_where();
    return s;
}

std::vector<Predicate> Parser::parse_where() {
    expect_keyword("WHERE");
    std::vector<Predicate> preds;
    do {
        preds.push_back(parse_predicate());
    } while (accept_keyword("AND"));
    return preds;
}

Predicate Parser::parse_predicate() {
    Predicate p;
    p.left = parse_column_ref();
    if (peek().kind != TokKind::SYMBOL)
        throw SQLException("expected comparison operator in condition");
    p.op = symbol_to_op(advance().text);
    // Right side: a literal value, or another column reference (joins).
    if (peek().kind == TokKind::NUMBER || peek().kind == TokKind::STRING ||
        is_symbol("-")) {
        p.right_is_column = false;
        p.right_value = parse_value();
    } else {
        p.right_is_column = true;
        p.right_col = parse_column_ref();
    }
    return p;
}

ColumnRef Parser::parse_column_ref() {
    ColumnRef ref;
    std::string first = expect_ident();
    if (accept_symbol(".")) {
        ref.table = first;
        ref.col = expect_ident();
    } else {
        ref.col = first;
    }
    return ref;
}

Value Parser::parse_value() {
    bool negative = accept_symbol("-");
    if (peek().kind == TokKind::NUMBER) {
        long long v = advance().number;
        return Value::make_int(negative ? -v : v);
    }
    if (peek().kind == TokKind::STRING) {
        if (negative) throw SQLException("cannot negate a string literal");
        return Value::make_text(advance().text);
    }
    throw SQLException("expected a literal value");
}

}  // namespace minidb
